pennylane/devices/default_qubit.py

```diff
@@ -552,80 +552,80 @@ def expval(self, observable, shot_range=None, bin_size=None):
-                    )
-                    c = qml.math.convert_like(coeff, product)
-
-                    if interface == "tensorflow":
-                        c = qml.math.cast(c, "complex128")
-
-                    res = qml.math.convert_like(res, product) + qml.math.sum(c * product)
+        if observable.name not in ("Hamiltonian", "SparseHamiltonian"):
+            return super().expval(observable, shot_range=shot_range, bin_size=bin_size)
 
+        assert self.shots is None, f"{observable.name} must be used with shots=None"
+
+        self.map_wires(observable.wires)
+        backprop_mode = (
+            not isinstance(self.state, np.ndarray)
+            or any(not isinstance(d, (float, np.ndarray)) for d in observable.data)
+        ) and observable.name == "Hamiltonian"
+
+        if backprop_mode:
+            # TODO[dwierichs]: This branch is not adapted to broadcasting yet
+            if self._ndim(self.state) == 2:
+                raise NotImplementedError(
+                    "Expectation values of Hamiltonians for interface!=None are "
+                    "not supported together with parameter broadcasting yet"
+                )
+            # We must compute the expectation value assuming that the Hamiltonian
+            # coefficients *and* the quantum states are tensor objects.
+
+            # Compute  <psi| H |psi> via sum_i coeff_i * <psi| PauliWord |psi> using a sparse
+            # representation of the Pauliword
+            res = qml.math.cast(qml.math.convert_like(0.0, observable.data), dtype=complex)
+            interface = qml.math.get_interface(self.state)
+
+            # Note: it is important that we use the Hamiltonian's data and not the coeffs
+            # attribute. This is because the .data attribute may be 'unwrapped' as required by
+            # the interfaces, whereas the .coeff attribute will always be the same input dtype
+            # that the user provided.
+            for op, coeff in zip(observable.ops, observable.data):
+                # extract a scipy.sparse.coo_matrix representation of this Pauli word
+                coo = qml.operation.Tensor(op).sparse_matrix(wires=self.wires, format="coo")
+                Hmat = qml.math.cast(qml.math.convert_like(coo.data, self.state), self.C_DTYPE)
+
+                product = (
+                    self._gather(self._conj(self.state), coo.row)
+                    * Hmat
+                    * self._gather(self.state, coo.col)
+                )
+                c = qml.math.convert_like(coeff, product)
+
+                if interface == "tensorflow":
+                    c = qml.math.cast(c, "complex128")
+
+                res = qml.math.convert_like(res, product) + qml.math.sum(c * product)
+
+        else:
+            # Coefficients and the state are not trainable, we can be more
+            # efficient in how we compute the Hamiltonian sparse matrix.
+            if observable.name in {"Hamiltonian", "SparseHamiltonian"}:
```

Comment:
Is it necessary? I think at this point `observable.name in {"Hamiltonian", "SparseHamiltonian"}:`

Reply:
haha I noticed that too, but decided to leave it. to clarify, the only change I did was indentation, and changing the if-condition to have a `not` while swapping the code being run.

I could change it here, but I left it because I figured the person who added it was scared that more observables would be added to the initial default-to-super list. if we're not worried, I can delete the conditional check too.

Reply:
Yes I would try to remove it, I don't think it will break anything. You can give it a try

Reply:
definitely won't break anything _today_. sounds good, I'll remove it though. changing that first line would probably make more of a mess than I can ever imagine anyway!

Reply:
lol just realized the code will actually crash if for some reason that statement _doesn't_ run, because then `Hmat` won't be defined!